### ISSUE TITLE
Help understanding GPU memory spikes during GRPO training.

### ISSUE THREAD

**Issue by ragingpandas** [volcengine/verl#456]
Hello, I'm currently experimenting with the performance of various configurations of GRPO to run on H200 gpus (2 nodes, 8 H200s per node). Regardless of my configuration I'm seeing large, intermittent memory spikes across some of the GPUs in a node. I was wondering if someone could provide some insight on what could be causing these large spikes (there also also more training job steps than spikes so it's not necessarily just every backward pass of the model, in the screenshot below the job had 29 steps).

I've attached an image of the GPU Memory and my config:

![Image](https://github.com/user-attachments/assets/87d7b838-1e1c-4266-a954-3dd9c9d72c55)

![Image](https://github.com/user-attachments/assets/844e8813-fa26-4d4d-a6e8-36901feb4f57)

```
#!/bin/bash

BASE_MODEL="Qwen/Qwen2.5-3B-Instruct"
ROLLOUT_TP_SIZE=1
EXPERIMENT_NAME="countdown-qwen2.5-3b-16k"
WANDB_PROJECT="RLCountdown"
DATA_DIR="countdown/dataset"

source scripts/base_script.sh "$@"

python3 -m trainer.main_ppo \
algorithm.adv_estimator=grpo \
data.train_files=$DATA_DIR/train.parquet \
data.val_files=$DATA_DIR/test.parquet \
data.train_batch_size=1024 \
data.max_prompt_length=256 \
data.max_response_length=16384 \
actor_rollout_ref.model.path=$BASE_MODEL \
actor_rollout_ref.actor.optim.lr=1e-6 \
actor_rollout_ref.model.use_remove_padding=True \
actor_rollout_ref.actor.ppo_mini_batch_size=256 \
actor_rollout_ref.actor.use_dynamic_bsz=True \
actor_rollout_ref.actor.ppo_max_token_len_per_gpu=35000 \
actor_rollout_ref.actor.use_kl_loss=True \
actor_rollout_ref.actor.kl_loss_coef=0.001 \
actor_rollout_ref.actor.kl_loss_type=low_var_kl \
actor_rollout_ref.model.enable_gradient_checkpointing=True \
actor_rollout_ref.actor.fsdp_config.param_offload=False \
actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
actor_rollout_ref.rollout.tensor_model_parallel_size=$ROLLOUT_TP_SIZE \
actor_rollout_ref.rollout.name=vllm \
actor_rollout_ref.rollout.gpu_memory_utilization=0.2 \
actor_rollout_ref.rollout.n=5 \
actor_rollout_ref.ref.fsdp_config.param_offload=True \
algorithm.kl_ctrl.kl_coef=0.001 \
trainer.critic_warmup=0 \
trainer.logger=['wandb'] \
trainer.experiment_name=$FULL_EXPERIMENT_NAME \
+trainer.val_before_train=True \
+trainer.raw_chat_template=False \
trainer.n_gpus_per_node=$N_GPUS \
trainer.nnodes=2 \
trainer.save_freq=-1 \
trainer.test_freq=20 \
trainer.project_name=$WANDB_PROJECT \
actor_rollout_ref.rollout.disable_log_stats=False \
trainer.total_epochs=15 2>&1 | tee verl_demo.log
```

**Comment by vermouth1992** [comment]
The spike is caused by cross entropy and entropy computation in backward.

**Comment by ragingpandas** [comment]
> The spike is caused by cross entropy and entropy computation in backward.

Thank you for the reply. Do you have any suggestions on what parameters I have set that causes the cross entropy and entropy memory usage during the backward pass to spike so much? 

Lowering the rollout gpu memory utilization, increasing the rollout TP, and lowering the ppo_max_token_len_per_gpu all still show these spikes and eventually result in a CUDA oom error. Or is this memory graph expected?

**Comment by PeterSH6** [comment]
I think you should lower the `ppo_max_token_len_per_gpu`. The KVCache and rollout weight are offloaded during training, so they won't consume GPU memory at the backward stage.

**Comment by ragingpandas** [comment]
I see, thanks for your reply. I halved the `ppo_max_token_len_per_gpu` (to 17k) and the spikes reduced down to 91% of GPU memory. Given vllm rollout memory isn't a constraint I'm guessing the recommendation would be to increase `actor_rollout_ref.rollout.log_prob_max_token_len_per_gpu` instead of just setting it to the same value as `actor_rollout_ref.actor.ppo_max_token_len_per_gpu`? Also if I'd like to scale up ppo_max_token_len_per_gpu for longer response length (or run with the same value on a larger model) what is your suggestion, when I tried this same config except with 4 nodes I still saw memory spikes and a CUDA OOM.

**Comment by kaamosi** [comment]
I encountered the same problem. The GPU memory usage shows spikes and CUDA OOM, any solutions?

**Comment by HorHang** [comment]
I encountered the same problems. Adjusting GPU utilization, i.e. `actor_rollout_ref.rollout.gpu_memory_utilization=0.8`, solved my case.

**Comment by zizi0123** [comment]
I encountered the same problem. I tried decrease gpu `actor_rollout_ref.rollout.gpu_memory_utilization` and decrease `actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu `, but they were not working..

**Comment by Purewhiter** [comment]
I have encountered the same problem, and not all actor updates will cause this. When memory spikes occured, the prompt and response length may not be very high compared to that when it didn't occur. Do you have any solutions?

**Comment by Kirrito-k423** [comment]
中文：感谢大家补充信息。根据前面的维护者回复，这类峰值主要来自 actor backward 阶段的 cross entropy / entropy 相关计算，而不是 rollout 阶段的 KV cache；`actor_rollout_ref.actor.ppo_max_token_len_per_gpu` 控制 `update_policy` 的 fwd/bwd token 上限，和 `actor_rollout_ref.ref/rollout.log_prob_max_token_len_per_gpu` 这种 forward-only log-prob 上限不是同一类预算。当前 perf tuning 文档也补充了 logits entropy 的显存说明：logits 通常是 `[bsz*seq_len, vocab]`，会产生较高峰值；可以尝试显式开启 `actor_rollout_ref.actor.entropy_checkpointing=True`，并在 ref/log-prob 侧尝试 `actor_rollout_ref.ref.entropy_from_logits_with_chunking=True`（以及合适的 `entropy_from_logits_chunk_size`）。同时请继续降低 actor 的 `ppo_max_token_len_per_gpu` 或使用 sequence parallel / activation offload 来确认峰值是否随 actor fwd/bwd token budget 下降。

为了判断这里是否还有 bug，能否请遇到 OOM 的用户补充：1) verl commit/version；2) 完整训练命令和覆盖后的 config；3) OOM stack trace；4) `actor_rollout_ref.rollout.disable_log_stats=False` 后的 rollout/log-prob 统计；5) `ppo_max_token_len_per_gpu`、`log_prob_max_token_len_per_gpu`、`entropy_checkpointing`、`entropy_from_logits_with_chunking` 的具体取值；6) 出问题 step 的 prompt/response token 分布。这样更容易区分是预期的 logits/cross-entropy 峰值，还是某个配置路径没有按预期分块/重算。

English: Thanks for the additional reports. Based on the maintainer comments above, this spike is mainly from actor backward cross-entropy / entropy-related computation, not rollout KV cache. `actor_rollout_ref.actor.ppo_max_token_len_per_gpu` controls the token budget for `update_policy` forward/backward, while `actor_rollout_ref.ref/rollout.log_prob_max_token_len_per_gpu` is a forward-only log-prob budget. The current perf tuning docs also call out the logits entropy peak: logits are typically `[bsz*seq_len, vocab]`, so they can be memory-heavy. Please try explicitly enabling `actor_rollout_ref.actor.entropy_checkpointing=True`, and for the ref/log-prob side try `actor_rollout_ref.ref.entropy_from_logits_with_chunking=True` with an appropriate `entropy_from_logits_chunk_size`. Also continue reducing actor `ppo_max_token_len_per_gpu`, or use sequence parallel / activation offload, to verify whether the peak scales with the actor fwd/bwd token budget.

To tell whether there is still a bug here, could affected users share: 1) verl commit/version; 2) full training command and resolved config; 3) CUDA OOM stack trace; 4) rollout/log-prob stats with `actor_rollout_ref.rollout.disable_log_stats=False`; 5) exact values for `ppo_max_token_len_per_gpu`, `log_prob_max_token_len_per_gpu`, `entropy_checkpointing`, and `entropy_from_logits_with_chunking`; 6) prompt/response token distribution for the failing step. That should help distinguish expected logits/cross-entropy memory peaks from a config path that is not chunking or recomputing as intended.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
